### PR TITLE
Cache Serialize API

### PR DESCRIPTION
**Pull request by hamersaw**
# TL;DR
Implemented cache reservations reducing the amount of duplicate work for long running cachable tasks. If enabled, a task acquires a reservation which is released upon completed (after the results are written to the cache). All other identical instances (ie. same task and inputs) wait for the reservation to expire or the cached results to become available.

## Type
 - [x] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
RFC - https://github.com/flyteorg/flyte/pull/1461

## Tracking Issue
fixes https://github.com/flyteorg/flyte/issues/267
fixes https://github.com/flyteorg/flyte/issues/420
fixes https://github.com/flyteorg/flyte/issues/872

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/controller/nodes/task/pre_post_execution.go [hunk 1]
```diff
@@ -188,3 +229,33 @@ func (t *Handler) ValidateOutputAndCacheAdd(ctx context.Context, nodeID v1alpha1
...
+	if tk.Metadata.Discoverable && tk.Metadata.DiscoverySerializable {
+		logger.Infof(ctx, "Catalog CacheSerializeEnabled: releasing catalog reservation.")
+		key := catalog.Key{
+			Identifier:     *tk.Id,
+			CacheVersion:   tk.Metadata.DiscoveryVersion,
+			TypedInterface: *tk.Interface,
+			InputReader:    inputReader,
+		}
+
+		err := t.catalog.ReleaseReservation(ctx, key, ownerID)
```
Comment (kumare3) [inline pkg/controller/nodes/task/pre_post_execution.go:249]: should we release the reservation on failure too?

File: pkg/controller/nodes/task/handler.go [hunk 2]
```diff
@@ -460,6 +476,13 @@ func (t Handler) invokePlugin(ctx context.Context, p pluginCore.Plugin, tCtx *ta
 		} else {
 			pluginTrns.ObserveSuccess(tCtx.ow.GetOutputPath(), &event.TaskNodeMetadata{CacheStatus: cacheStatus.GetCacheStatus(), CatalogKey: cacheStatus.GetMetadata()})
 		}
+
+		ownerID := tCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName()
+		reservation, err := t.ReleaseCatalogReservation(ctx, ownerID, tCtx.tr, tCtx.InputReader())
```
Comment (kumare3) [inline pkg/controller/nodes/task/handler.go:481]: ahh i see it, but should we do this in finalize?

Comment (kumare3) [reply pkg/controller/nodes/task/handler.go:481]: one problem is today we optimize finalize as it is a discrete step, we optimize it and use only for non pod tasks

Comment (hamersaw) [reply pkg/controller/nodes/task/handler.go]: done. this makes perfect sense, in addition to failures it releases the reservations on aborts as well.

File: go.mod [hunk 3]
```diff
@@ -33,3 +33,5 @@ require (
 )
 
 replace github.com/aws/amazon-sagemaker-operator-for-k8s => github.com/aws/amazon-sagemaker-operator-for-k8s v1.0.1-0.20210303003444-0fb33b1fd49d
+
+replace github.com/flyteorg/flyteplugins => github.com/hamersaw/flyteplugins v0.6.2-0.20211129132447-34fdd4b8118e
```
Comment (EngHabu) [inline go.mod:37]: Reminder to replace this

**Review by EngHabu — APPROVED** [review]

**Review by EngHabu — APPROVED** [review]
